### ISSUE TITLE
help with understanding the intuition behind the "4096" attention window.

### ISSUE THREAD

**Issue by vr25** [allenai/longformer#48]
Hi,

I understand that in order to overcome the limitation of BERT, Transformer-XL, etc with a longer attention window, LongFormer does it by pooling all the local attentions (512) together in global attention (512 x 8 = 4096). The selection of "4096" as the extended attention window isn't clear to me in the paper. 

My understanding of the above is the GPU memory training and evaluation constraints because all the 4096 or perhaps, just the 512 tokens are processed simultaneously. So, does it mean that this model can work beyond the 4096 attention window as well? Say, for very long legal documents with 100,000 tokens. In such cases, is it advisable to break it down to say, 100,000/512 = 196 (approx.) local blocks?

Also, the output is just a number associated with each long document (regression) or a class label (classification).

Any suggestions for this challenge will be helpful.

Thank you!

**Comment by ibeltagy** [comment]
> LongFormer does it by pooling all the local attentions (512) together in global attention (512 x 8 = 4096).

This is not entirely true. There's no "pooling" of the 4096 tokens into 512. We keep all 4096 tokens. The only change is how attention is computed; instead of every token attending to every other token, we change it such that every token attends to a smaller number of surrounding tokens. This speeds up selfattention computation (which is the bottleneck) by assuming that the attention score between certain pairs of words is zero. This doesn't change the architecture or introduce any pooling. 

We are working on some code that will make it easy to train your own long model, so you can try longer sequences. We know it is easy to get to 16K or even 32k with RoBERTa-base architecture (need base model, fp16, gradient checkpointing). For sequences longer than that, you will need to find ways to save memory depending on your application. For example, reducing window size, reducing size of the feed forward layers, implementing reversible transformers, use sinusoidal position embedding instead of learned position embedding.

**Comment by vr25** [comment]
Thank you for the explanation. I thought RoBERTa has an attention constraint of 512 tokens too [[here](https://github.com/huggingface/transformers/issues/1215)], just clarifying.

**Comment by ibeltagy** [comment]
RoBERTa has a 512 limit. Our pretrained model has a 4096 limit.

**Comment by ibeltagy** [comment]
Closing. Please feel free to reopen or create a new one if you have more questions.
